src/components/include/cbuf.h

```diff
@@ -405,241 +341,131 @@ __cbuf_alloc_meta_inconsistent(struct cbuf_alloc_desc *d, struct cbuf_meta *m)
+	do {
+		*old   = *target;
+		cm		   = CBUFM_GET_NEXT(fl);
+		new_head.next_flag = cm->next_flag;
+		new_head.cbid.tag  = fl->cbid.tag+1;
+	} while(unlikely(!cos_dcas(target, *old, *update)));
+	if (unlikely(cm == CBUFM_GET_NEXT(cm))) {
+		cm   = __cbuf_alloc_slow(sz, &len);
+		assert(cm);
 		goto done;
 		/*
 		 * TODO: check if this cbuf has been taken by another
 		 * thd already.  shall we add this cbuf to the
 		 * freelist and just continue?
-		 */
-	} 
-	REM_LIST(d, next, prev);
-	assert(EMPTY_LIST(d, next, prev));
-	cbid  = d->cbid;
-	assert(cbid);
-	cbidx            = cbid_to_meta_idx(cbid);
-	cm               = cbuf_vect_lookup_addr(cbidx, tmem);
-
-	mapped_in        = cbufm_is_mapped(cm);
-	already_used     = cm->nfo.c.refcnt;
-	flags            = CBUFM_TOUCHED;
-	if (tmem) flags |= CBUFM_TMEM;
-	cm->nfo.c.flags |= flags;
-	if(cm->nfo.c.refcnt == CBUFP_REFCNT_MAX)
-		assert(0);
-	cm->nfo.c.refcnt++;
-
-	/* 
-	 * Now that IN_USE is set, we know the manager will not rip
-	 * this out from under us.  Check that nothing has changed,
-	 * and the pointer is consistent with the allocation
-	 * descriptor 
-	 */
+		 */   //???
+	}
+	already_used = CBUFM_GET_REFCNT(cm);
 	assert(!already_used);
-	if (unlikely(__cbuf_alloc_meta_inconsistent(d, cm) || already_used || !mapped_in)) {
-		/* 
-		 * This is complicated.
-		 *
-		 * See cbuf_slab_free for the rest of the story.  
-		 *
-		 * Assumptions: 
-		 * 
-		 * 1) The cbuf manager shared the cbuf meta (in the
-		 * meta_cbuf vector) information with this component.
-		 * It can remove asynchronously a cbuf from this
-		 * structure at any time IFF that cbuf is marked as
-		 * ~CBUF_IN_USE.
-		 *
-		 * 2) The slab descriptors, and the slab_desc vector
-		 * are _not_ shared with the cbuf manager for
-		 * complexity reasons.
-		 *
-		 * Question: How do we reconcile the fact that the
-		 * cbuf mgr might remove at any point a cbuf from this
-		 * component, but we still have a slab descriptor
-		 * lying around for it?  How will we know that the
-		 * cbuf has been removed, and not to use the slab
-		 * data-structure anymore?
-		 *
-		 * Answer: The slabs are deallocated lazily (seen
-		 * here).  When a slab is pulled off of the freelist
-		 * (see the following code), we check to make sure
-		 * that the cbuf meta information matches up with the
-		 * slab's information (i.e. the cbuf id and the
-		 * address in memory of the cbuf.  If they do not,
-		 * then we know that the slab is outdated and that the
-		 * cbuf backing it has been taken from this component.
-		 * In that case (shown in the following code), we
-		 * delete the slab descriptor.  Again, see
-		 * cbuf_slab_free to see the surprising fact that we
-		 * do _not_ deallocate the slab descriptor there to
-		 * reinforce that point.
-		 */
-		__cbuf_desc_free(d);
-		if (likely(!already_used)) {
-			cm->nfo.c.flags &= ~CBUFM_TOUCHED;
-			cm->nfo.c.refcnt--;
-		}
+	inconsistent = CBUF_INCONSISENT(cm);
```

Comment:
We need to review the inconsistency rules, and you need to document them in a comment somewhere.  I can't review that here.